### PR TITLE
Promote mathematical functions to `Kokkos` namespace

### PR DESCRIPTION
**Pull request by dalg24**
Move all math functions from `Kokkos::Experimental::` to `Kokkos::` namespace.
Provide using-declaration in the `Kokkos::Experimental::` namespace when deprecated code is enabled to ease the transition.

Deprecated function `log2(unsigned) -> int` is being removed.

### CODE REVIEW

File: core/src/impl/Kokkos_BitOps.hpp [hunk 1]
```diff
@@ -152,15 +152,6 @@ unsigned integral_power_of_two_that_contains(const unsigned N) {
...
 
 }  // namespace Impl
-
-#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_3
-
-KOKKOS_DEPRECATED KOKKOS_INLINE_FUNCTION int log2(unsigned i) {
-  return Impl::int_log2(i);
-}
-
-#endif
```
Comment (dalg24) [inline core/src/impl/Kokkos_BitOps.hpp:162]: We can't keep it if we proceed and promote the math functions.

File: core/src/Kokkos_MathematicalFunctions.hpp [hunk 2]
```diff
@@ -99,6 +101,16 @@ namespace Experimental {
 #endif
 #endif
 
+#if defined(KOKKOS_ENABLE_DEPRECATED_CODE_3)
+#define KOKKOS_IMPL_MATH_FUNCTIONS_DEFINED_IF_DEPRECATE_CODE_ENABLED( \
```
Comment (masterleinad) [inline core/src/Kokkos_MathematicalFunctions.hpp:105]:
```suggestion
#define KOKKOS_IMPL_MATH_FUNCTIONS_DEFINED_IF_DEPRECATED_CODE_ENABLED( \
```

**Review by masterleinad — APPROVED** [review]
`KOKKOS_IMPL_MATH_FUNCTIONS_DEFINED_IF_DEPRECATED_CODE_ENABLED` makes more sense to me than `KOKKOS_IMPL_MATH_FUNCTIONS_DEFINED_IF_DEPRECATE_CODE_ENABLED` but I don't think it matters much.

File: core/src/Kokkos_MathematicalFunctions.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/Kokkos_MathematicalFunctions.hpp:105]: You are right.  It is a typo.  Will fix.

File: core/src/Kokkos_MathematicalFunctions.hpp [hunk 3]
```diff
@@ -47,6 +47,7 @@
 
 #include <Kokkos_Macros.hpp>
 #include <cmath>
+#include <cstdlib>
```
Comment (Rombur) [inline core/src/Kokkos_MathematicalFunctions.hpp:50]: Why did you add this include?

File: core/src/Kokkos_MathematicalFunctions.hpp [hunk 4]
```diff
@@ -76,16 +77,17 @@ struct promote<float> {
 };
 template <class T>
 using promote_t = typename promote<T>::type;
-template <class T, class U>
+template <class T, class U,
+          bool = std::is_arithmetic<T>::value&& std::is_arithmetic<U>::value>
```
Comment (Rombur) [inline core/src/Kokkos_MathematicalFunctions.hpp:81]:
```suggestion
          bool = std::is_arithmetic<T>::value && std::is_arithmetic<U>::value>
```

**Review by Rombur — APPROVED** [review]

File: core/src/Kokkos_MathematicalFunctions.hpp (hunk 4 above)
Comment (dalg24) [reply core/src/Kokkos_MathematicalFunctions.hpp:81]: I also dislike it but that is how Clang-Format wants it.

File: core/src/Kokkos_MathematicalFunctions.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/Kokkos_MathematicalFunctions.hpp]: Rebased and applied your suggestion

**Review by crtrott — APPROVED** [review]
